### ISSUE TITLE
[RFE] Add ability to modify already set module parameters

### ISSUE THREAD

**Issue by Cropi** [rsyslog/rsyslog#4978]
### Justification and typical use case
On RHEL/Fedora Linux distributions, we do ship rsyslog with some basic configuration file(```/etc/rsyslog.conf```) where we specify the most important modules and other directives. For example we specify the default template for omfile actions via ```module(load="builtin:omfile" Template="RSYSLOG_TraditionalFileFormat")```
We do not recommend(allow) users to alter this file, instead we suggest to add new configuration snippets inside an included directory(```/etc/rsyslog.d```). However, some users want e.g. to alter the default template for the omfile module but as far as I know it's not possible via the RainerScript syntax. Another use case could be, for example we use ansible to install rsyslog on various systems and by default the basic configuration file is also installed. Thus, we can not alter the default template of the omfile module, instead we need to rewrite the entire config file.

### Side note
It was doable with the legacy syntax:
```
$ActionFileDefaultTemplate SomeCustomTemplate
```

### Possible approach
Introduce new configuration directive, e.g.:
```
# /etc/rsyslog.conf
module(load="builtin:omfile" Template="RSYSLOG_TraditionalFileFormat")
```
```
# /etc/rsyslog.d/00-default-changes.conf
moduleChange(type="omfile" template="newTemplate")
```

I know it would require code changes in other modules as well but I think it's worth the discussion. At the beginning only a few modules would support this option. What do you think? I do not want to jump into implementation details before the idea is clear and accepted.

**Comment by davidelang** [comment]
using the action() syntax, the intent is that everything is defined in that 
statement, and you don't need to look elsewhere in the config to figure out what 
it's going to do

the older syntax of $foo=bar followed by an output line results in configs that 
don't do what people think they do.

adding back the ability to change the behavior of later lines (especially via 
include files) doesn't seem like a great idea

with included files, people seem to think that the things in one file only 
affect outputs in that file, they don't seem to understand that all the files 
get combined and evaluated as if they were one big file.

I also really disagree with the idea that the default configs should not be 
changed, there are a LOT of things that are default just because that's how 
syslog was configured back in the 90's when all logs were reviewed manually that 
have far less significance today when approximatly 0% of logs are reviewed 
manually. There are also things like pstats that need to be configured in the 
file, and the last I checked, even the redhat default file had commented out 
portions for people to uncomment to enable specific things.

Yes, I sometimes wish that I could set defaults instead of listing them 
explicitly in every action() statement, but then I help troubleshoot the old 
style configs where $foo lines change (or fail to change) the behavior of other 
outputs in ways that the admin doesn't expect, and am reminded of the tradeoff 
involved here.

David Lang

**Comment by rgerhards** [comment]
> adding back the ability to change the behavior of later lines (especially via include files) doesn't seem like a great idea with included files, people seem to think that the things in one file only affect outputs in that file, they don't seem to understand that all the files get combined and evaluated as if they were one big file.

I fully agree with @davidelang here. The main reason we introduced the new script format was to disallow changing default at arbitrary and unexpected places. That made writing configs and troubleshooting very hard.

Nevertheless, if you still need that functionality, you can use legacy statements. They are perfectly fine and keep being supported. Using them also makes clear which kinds of problems that involves.

In that sense, we do not need to change config processing. The requested feature is still present. And this also shows why it is problematic and was abandoned.

*If I may suggest an alternate route for Red Hat:*

* create a file default-omfile-fmt.conf which just has the default template setting
* include that at the very beginning of rsyslog.conf

If I understood you philosophy right, that would enable users to change the default template to their needs without touching the main rsyslog.conf. It would also make crystal-clear that there indeed is only one default.

**Comment by Cropi** [comment]
Thank you for the detailed answers. I understand the reason behind the new syntax, so it's easier to know what parameters belong to which configuration directives. In this way, it's less likely that users will mix up the configuration.
We will either go with the legacy syntax or use the alternate route as suggested by Rainer.

Thanks.
